In [1]:
import pandas as pd
import numpy as np

In [3]:
import pandas as pd

# Reading the CSV files
d1 = pd.read_csv("student-mat.csv", sep=";")
d2 = pd.read_csv("student-por.csv", sep=";")

# Merging the DataFrames on specified columns
merge_columns = ["school", "sex", "age", "address", "famsize", "Pstatus", "Medu", "Fedu", "Mjob", "Fjob", "reason", "nursery", "internet"]
d3 = pd.merge(d1, d2, on=merge_columns)

# Print the number of rows in the merged DataFrame
print(f"Number of students: {len(d3)}")
d3.head()

Number of students: 382


school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob  ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher  ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other  ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other  ...   
3     GP   F   15       U     GT3       T     4     2   health  services  ...   
4     GP   F   16       U     GT3       T     3     3    other     other  ...   

  famrel_y freetime_y  goout_y  Dalc_y  Walc_y health_y absences_y G1_y G2_y  \
0        4          3        4       1       1        3          4    0   11   
1        5          3        3       1       1        3          2    9   11   
2        4          3        2       2       3        3          6   12   13   
3        3          2        2       1       1        5          0   14   14   
4        4          3        2       1       2        5          0   11   13   

  G3_y  
0   11  
1   11  
2   12  
3   14  
4   13  

[5 rows x 53 columns]

In [4]:
print(d3.columns)

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian_x', 'traveltime_x', 'studytime_x',
       'failures_x', 'schoolsup_x', 'famsup_x', 'paid_x', 'activities_x',
       'nursery', 'higher_x', 'internet', 'romantic_x', 'famrel_x',
       'freetime_x', 'goout_x', 'Dalc_x', 'Walc_x', 'health_x', 'absences_x',
       'G1_x', 'G2_x', 'G3_x', 'guardian_y', 'traveltime_y', 'studytime_y',
       'failures_y', 'schoolsup_y', 'famsup_y', 'paid_y', 'activities_y',
       'higher_y', 'romantic_y', 'famrel_y', 'freetime_y', 'goout_y', 'Dalc_y',
       'Walc_y', 'health_y', 'absences_y', 'G1_y', 'G2_y', 'G3_y'],
      dtype='object')


Formulating the Hypothesis

Null Hypothesis (H0): There is no significant correlation between the number of school absences and student final grades.

Alternative Hypothesis (H1): There is a significant correlation between the number of school absences and student final grades.

In [5]:
from scipy.stats import pearsonr, spearmanr

# Calculate Pearson's correlation for normally distributed data or linear relationships
pearson_corr_math, p_value_pearson_math = pearsonr(d3['absences_x'], d3['G3_x'])

# Calculate Spearman's correlation for non-parametric data or non-linear relationships
spearman_corr_math, p_value_spearman_math = spearmanr(d3['absences_x'], d3['G3_x'])

print(f"Pearson's correlation (Math): {pearson_corr_math:.3f}, p-value: {p_value_pearson_math:.3f}")
print(f"Spearman's correlation (Math): {spearman_corr_math:.3f}, p-value: {p_value_spearman_math:.3f}")


Pearson's correlation (Math): 0.029, p-value: 0.572
Spearman's correlation (Math): 0.015, p-value: 0.764


In [6]:
# Calculate Pearson's correlation for normally distributed data or linear relationships
pearson_corr_portuguese, p_value_pearson_portuguese = pearsonr(d3['absences_y'], d3['G3_y'])

# Calculate Spearman's correlation for non-parametric data or non-linear relationships
spearman_corr_portuguese, p_value_spearman_portuguese = spearmanr(d3['absences_y'], d3['G3_y'])

print(f"Pearson's correlation (Portuguese): {pearson_corr_portuguese:.3f}, p-value: {p_value_pearson_portuguese:.3f}")
print(f"Spearman's correlation (Portuguese): {spearman_corr_portuguese:.3f}, p-value: {p_value_spearman_portuguese:.3f}")


Pearson's correlation (Portuguese): -0.145, p-value: 0.005
Spearman's correlation (Portuguese): -0.176, p-value: 0.001


Math Subject Correlation Results:
Pearson's correlation coefficient is 0.029 with a p-value of 0.572.
Spearman's correlation coefficient is 0.015 with a p-value of 0.764.
For the Math subject, both Pearson's and Spearman's correlation coefficients are very close to zero, indicating a very weak linear and rank correlation between school absences and final grades. Additionally, the p-values are well above the common significance level of 0.05, suggesting that any correlation observed is not statistically significant. This implies that, for the Math subject, there's no evidence to suggest that the number of absences significantly impacts the final grades.

Portuguese Subject Correlation Results:
Pearson's correlation coefficient is -0.145 with a p-value of 0.005.
Spearman's correlation coefficient is -0.176 with a p-value of 0.001.
For the Portuguese subject, both Pearson's and Spearman's correlation coefficients are negative, indicating a weak but negative relationship between school absences and final grades – as absences increase, grades tend to decrease. The p-values for both tests are below the significance level of 0.05, indicating that this correlation is statistically significant, although the relationship is relatively weak.

Interpretation and Next Steps:
Math: There's no significant evidence to suggest that absences impact final grades. This might imply that factors other than attendance play a more critical role in determining Math performance.

Portuguese: The significant negative correlation, although weak, suggests that increased absences may be associated with lower grades. This relationship warrants further investigation, perhaps exploring if there are thresholds of absences that are particularly detrimental, or if other factors (like reasons for absences) might mediate this relationship.

Given the significant findings for Portuguese, educators and policymakers might consider interventions to reduce absences or provide additional support to students who miss school. However, given the weak nature of the correlation, it's also crucial to investigate other variables that might have a more substantial impact on academic performance.